### DISCUSSION TITLE
React.cloneElement and MobX obsever HOC

### DISCUSSION THREAD

**Discussion by jamieathans** [mobxjs/mobx#3398] [Q&A] [answered]
Greetings,

Are there known incompatibilities between calling `React.cloneElement` on components wrapped in MobX `observer`?

The error appears to manifest as:

```
Uncaught Error: React.cloneElement(...): The argument must be a React element, but you passed undefined
```

I've introduced MobX into a legacy React code base and it's working well except for the above issue where `React.cloneElement` is used for adding props from parent to children components.

**Comment by mweststrate** [comment]
cloneElement clones elements, not components, so I don't think it should
normally matter in any way.

**Reply by jamieathans** [reply]
@mweststrate Just to verify, this is what is breaking in my scenario:

```
const MyObservedChildComponent = observer(function MyObservedChildComponent() {
...
});
```

and in some parent component,

```
React.cloneElement(props.children, {...});
```

where `props.children` is `MyObservedChildComponent`.

You're saying the above scenario should be OK?

**Reply by urugator** [reply]
> where props.children is MyObservedChildComponent

Is the child passed like this:
`<SomeParent>{MyObservedChildComponent}</SomeParent>`
or like this:
`<SomeParent><MyObservedChildComponent /></SomeParent>`
?
The first passes a component (function), which won't work with `React.cloneElement`, the second passes an element (object), which should work with `React.cloneElement`.

**Reply by jamieathans** [reply]
It's existing code using React Router 3 nested routing and the child is passed like first method

` <Route path=":applicationId" component={ApplicationRoute}>`

If `ApplicationRoute` is wrapped in `observer` things break - as soon as I remove the `observer` HOC things start working again.

**Reply by urugator** [reply]
Well you definitely can't pass component (observable or not) to `React.cloneElement`, so it's probably not about `cloneElement` unless you use it somehow inappropriately.

**Reply by jamieathans** [reply]
So I made the error happen again by applying `observer` to one of the components in the route chain and this is the stack trace:


```
Warning: React.createElement: type is invalid -- expected a string (for built-in components) or a class/function (for composite components) but got: symbol.

Check the render method of `RouterContext`.
    in RouterContext (created by Router)
    in Router react_devtools_backend.js:3973:25
    overrideMethod react_devtools_backend.js:3973
    React 3
    createElement RouterContext.js:49
    element RouterContext.js:94
    render RouterContext.js:64
    React 8
    unstable_runWithPriority scheduler.development.js:653
    React 26
    js route.js:154
    Webpack 7

Uncaught Error: React.cloneElement(...): The argument must be a React element, but you passed undefined.
    React 2
    render app.js:196
    React 11
    unstable_runWithPriority scheduler.development.js:653
    React 26
    js route.js:154
    Webpack 7
react.development.js:820
    React 2
    render app.js:196
    React 10
    performSyncWorkOnRoot self-hosted:1168
    flushSyncCallbackQueueImpl React
    unstable_runWithPriority scheduler.development.js:653
    React 6
    bind_applyFunctionN self-hosted:1319
    dispatchDiscreteEvent self-hosted:1282
    React 20
    js route.js:154
    Webpack 7

The above error occurred in the <App> component:
    in App (created by AppWrapper)
    in AppWrapper (created by RouterContext)
    in RouterContext (created by Router)
    in Router

Consider adding an error boundary to your tree to customize error handling behavior.
Visit https://fb.me/react-error-boundaries to learn more about error boundaries. react_devtools_backend.js:3973:25
    overrideMethod react_devtools_backend.js:3973
    React 11
    unstable_runWithPriority scheduler.development.js:653
    React 5
    unstable_runWithPriority scheduler.development.js:653
    React 26
    js route.js:154
    Webpack 7

Uncaught Error: React.cloneElement(...): The argument must be a React element, but you passed undefined.
    React 2
    render app.js:196
    React 11
    unstable_runWithPriority scheduler.development.js:653
    React 26
    js route.js:154
    Webpack 7
        __webpack_require__
        fn
        0
        __webpack_require__
        checkDeferredModules
        <anonymous>
        <anonymous>
react.development.js:820
    React 2
    render app.js:196
    React 10
    performSyncWorkOnRoot self-hosted:1168
    flushSyncCallbackQueueImpl React
    unstable_runWithPriority scheduler.development.js:653
    React 6
    bind_applyFunctionN self-hosted:1319
    dispatchDiscreteEvent self-hosted:1282
    React 20
    js route.js:154
    Webpack 7
```

**Reply by urugator** [reply]
Just a shot in the dark, but try to replace `observer(MyCmp)` with  `React.memo(MyCmp)` and check if it throws

**Reply by urugator** [reply]
Also try to replace the component with a simple/dummy verssion like `observer(() => "ok")`

**Reply by jamieathans** [reply]
> Just a shot in the dark, but try to replace `observer(MyCmp)` with `React.memo(MyCmp)` and check if it throws

Well there you go - replacing `observer(MyCmp)` with `React.memo(MyCmp)` throws as well.

So `React.memo` is the issue?

**Reply by jamieathans** [reply]
The workaround I've come up with is to create an intermediate non observed Component:

```
const ObservedComponent = observer(function ObservedComponent(props) {
...
});

function NonObservedComponent(props) {
    return (
        <ObservedComponent {...props} />
    );
}

export default NonObservedComponent;
```

**Comment by jamieathans** [comment, marked as answer]
As per the suggestion by @urugator, adding `React.memo` breaks things as well, so the issue is not specific to the MobX `observer` HOC.

**Reply by mweststrate** [reply]
`React.cloneElement(props.children)` is a dangerous pattern anyway, as it will break on non singular children, make sure to at least use the `React.onlyChild` react api for that (or something similar, forgot the exact name)
